### PR TITLE
[Feature] Multithread memmap

### PR DESCRIPTION
**Pull request by vmoens**
TODO:
- [ ] Doc
- [x] Tests
- [x] hide the executor logit under a private `_memmap_`
- [x] Make a `memmap` version that does not change the tenosrdict inplace

### CODE REVIEW

File: tensordict/_td.py [hunk 1]
```diff
@@ -1510,21 +1518,26 @@ def save_metadata(data: TensorDictBase, filepath, metadata=None):
...
+                        filename=str(prefix / f"{key}.memmap")
+                        if prefix is not None
+                        else None,
+                        copy_existing=copy_existing,
+                        # copy_existing=copy_existing,
+                    )
+                if executor is None:
+                    _populate()
+                else:
+                    executor.submit(_populate)
```
Comment (laurencer) [inline tensordict/_td.py:1540]: Probably want to return a Future here so the caller can block on the saves actually completing if they want (or get a callback when it completes)

**Comment by vmoens** [comment]
@laurencer thanks for the review.
I created the futures (execution time is roughly identical as before). I thought that `executor.__exit__()` would wait for submitted jobs to complete but after testing it, it doesn't seem to be the case. Do we want to optionally let users proceed with the main script without waiting for the futures to complete (and return the futures)?

I also prevented conflicting executor and num_threads, but eventually I think `executor` will only be an arg of a sub-call to a new _memmap private function (as well as `futures` to restrict the args of the public method to what they should be).

**Comment by fegin** [comment]
@vmoens Thanks for the information. Just out of curiosity, have you try to compare the performance if the target storage is a memory file system?

**Comment by vmoens** [comment]
That would be `memmap_()` (without file associated). I can check that!

**Pull request merged**